lib/pharos/cluster_manager.rb

```diff
@@ -40,30 +40,47 @@ def validate
       addon_manager.validate
     end
 
+    # @return [Pharos::Configuration::Host]
+    def preferred_master_host
+      master = config.master_hosts.find { |h|
+        ssh = ssh_manager.client_for(h)
+        ssh.file('/etc/kubernetes/kubelet.conf').exist?
+      }
+      if master
+        master
+      else
+        config.master_host
+      end
+    end
+
     def apply_phases
+      master = preferred_master_host
       apply_phase(Phases::ValidateHost, config.hosts, ssh: true, parallel: true)
       apply_phase(Phases::MigrateMaster, config.master_hosts, ssh: true, parallel: true)
       apply_phase(Phases::ConfigureHost, config.hosts, ssh: true, parallel: true)
-      apply_phase(Phases::ConfigureCfssl, config.etcd_hosts, ssh: true, parallel: true)
-      apply_phase(Phases::ConfigureEtcdCa, config.etcd_hosts[0...1], ssh: true, parallel: false)
-      apply_phase(Phases::ConfigureEtcd, config.etcd_hosts, ssh: true, parallel: true)
+      unless @config.etcd&.hosts
+        apply_phase(Phases::ConfigureCfssl, config.etcd_hosts, ssh: true, parallel: true)
+        apply_phase(Phases::ConfigureEtcdCa, config.etcd_hosts[0...1], ssh: true, parallel: false)
+        apply_phase(Phases::ConfigureEtcdChanges, config.etcd_hosts[0...1], ssh: true, parallel: false)
+        apply_phase(Phases::ConfigureEtcd, config.etcd_hosts, ssh: true, parallel: true)
+      end
 
       apply_phase(Phases::ConfigureSecretsEncryption, config.master_hosts, ssh: true, parallel: false)
       apply_phase(Phases::ConfigureMaster, config.master_hosts, ssh: true, parallel: false)
-      apply_phase(Phases::MigrateWorker, config.worker_hosts, ssh: true, parallel: true, master: config.master_host)
+      apply_phase(Phases::MigrateWorker, config.worker_hosts, ssh: true, parallel: true, master: master)
```

Comment:
This is going to bug out if the master changes... I already fixed this as part of #278, I'll split that out as a separate fix PR.

Reply:
See #302